src/Tables.jl

```diff
@@ -31,110 +31,178 @@ function __init__()
+        Tables.eachcolumn(sch, row, mytbl) do val, col, name, mytbl
+            push!(mytbl[col], val)
+        end
     end
     return mytbl
 end
 ```
 
-Alternatively, if `MyTableType` is column-oriented, perhaps my definition would be more like:
-
+Alternatively, if `MyTable` is column-oriented, perhaps my definition would be more like:
 ```julia
-function MyTableType(x)
+function MyTable(x)
+    Tables.istable(x) || throw(ArgumentError("MyTable requires a table input"))
     cols = Tables.columns(x)
-    return MyTableType(collect(map(String, keys(cols))), [col for col in cols])
+    # here we use Tables.eachcolumn to iterate over each column in a `Columns` object
+    return MyTable(collect(Tables.names(cols)), [collect(col) for col in Tables.eachcolumn(cols)])
 end
 ```
 
-Obviously every table type is different, but via a combination of `Tables.schema`, `Tables.rows`, and `Tables.columns`, each table type should be able to construct an instance of itself.
+Obviously every table type is different, but via a combination of `Tables.rows` and `Tables.columns` each table type should be able to construct an instance of itself.
 """
 abstract type Table end
 
-abstract type AccessStyle end
-struct RowAccess <: AccessStyle end
-struct ColumnAccess <: AccessStyle end
-
-"Tables.schema(s) => NamedTuple{names, types}"
-function schema end
+# default definitions
+rowaccess(x) = false
+columnaccess(x) = false
 
 function istable(::Type{T}) where {T}
-    hasmethod(Tables.AccessStyle, Tuple{T}) &&
-    hasmethod(Tables.schema, Tuple{T}) &&
-    (Tables.AccessStyle(T) === Tables.RowAccess() ? hasmethod(Tables.rows, Tuple{T}) :
-    Tables.AccessStyle(T) === Tables.ColumnAccess() ? hasmethod(Tables.columns, Tuple{T}) : false)
+    if hasmethod(Tables.rowaccess, Tuple{T})
+        return hasmethod(Tables.rows, Tuple{T})
+    elseif hasmethod(Tables.columnaccess, Tuple{T})
+        return hasmethod(Tables.columns, Tuple{T})
+    end
+    return false
+end
+
+# Schema implementation
+"""
+    Tables.Schema(names, types)
+
+Create a `Tables.Schema` object that holds the column names and types for a tabular data object.
+`Tables.Schema` is dual-purposed: provide an easy interface for users to query these properties,
+as well as provide a convenient "structural" type for code generation. 
+
+To access the names, one can simply call `sch.names` to return the tuple of Symbols.
+To access column types, one can similarly call `sch.types`, which will return either a tuple of types (like `(Int64, Float64, String)`)
+or `nothing` if the column types are not computable or otherwise unknown.
+
+The actual type definition is
+```julia
+struct Schema{names, types} end
+```
+Where `names` is a tuple of Symbols, and `types` is a tuple _type_ of types (like `Tuple{Int64, Float64, String}`).
+Encoding the names & types as type parameters allows convenient use of the type in generated functions
+and other optimization use-cases.
+"""
+struct Schema{names, types} end
```

Comment:
A small comment: current implementation of `Schema` differs from `NamedTuple` also because it allows duplicate column names (as opposed to `NamedTuple`). Not sure if it is a significant thing.